File: src/main/java/com/lera/autocomplete/services/StateService.java
package com.lera.autocomplete.services;

import com.lera.autocomplete.exception.StateAlreadyExistsException;
import com.lera.autocomplete.model.State;

import java.util.List;

public interface StateService {
    List<State> getAllStates();
    List<State> getAllStatesStartingWith(String query);
    State getState(Integer stateId) throws Exception;
    State createState(State state) throws StateAlreadyExistsException;
    State modifyState(Integer stateId, State state);
    void deleteState(Integer stateId);
}

File: src/main/frontend/src/app/services/state.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {State} from "../shared/state";

@Injectable({
  providedIn: 'root'
})
export class StateService {
  private url = "http://localhost:8080/api/states";

  constructor(private http: HttpClient) {
  }

  getAllStates(): Observable<State[]> {
    return this.http.get<State[]>(this.url);
  }

  getStatesStartingWith(query: string): Observable<State[]> {
    return this.http.get<State[]>(this.url + "/query/" + query);
  }

  createState(state: State): Observable<State> {
    return this.http.post<State>(this.url, state);
  }

  public modifyState(state: State): Observable<State> {
    return this.http.put<State>(this.url + '/' + state.stateId, state);
  }

  deleteState(stateId: number): Observable<any> {
    return this.http.delete(this.url + '/' + stateId);
  }
}
